tensorboard/db.py

```diff
@@ -384,7 +408,357 @@ def create_event_logs_table_path_index(self):
+
+    :rtype: int
+    """
+    self._check_that_read_query_was_issued()
+    return self._delegate.rowcount
+
+  @property
+  def lastrowid(self):
+    """Returns last row ID.
+
+    :rtype: int
+    """
+    self._check_that_read_query_was_issued()
+    return self._delegate.lastrowid
+
+  def _get_arraysize(self):
+    self._init_delegate()
+    return self._delegate.arraysize
+
+  def _set_arraysize(self, arraysize):
+    self._init_delegate()
+    self._delegate.arraysize = arraysize
+
+  arraysize = property(_get_arraysize, _set_arraysize)
+
+  def close(self):
+    """Closes resources associated with cursor."""
+    if self._delegate is not None:
+      self._delegate.close()
+      self._delegate = None
+    self._is_closed = True
+
+  def __iter__(self):
+    """Returns iterator over results of last read query.
+
+    :rtype: types.GeneratorType[tuple[object]]
+    """
+    self._check_that_read_query_was_issued()
+    for row in self._delegate:
+      yield row
+
+  def nextset(self):
+    """Raises NotImplementedError."""
+    raise NotImplementedError('Cursor.nextset not supported')
+
+  def callproc(self, procname, parameters=()):
+    """Raises NotImplementedError."""
+    raise NotImplementedError('Cursor.callproc not supported')
+
+  def setinputsizes(self, sizes):
+    """Raises NotImplementedError."""
+    raise NotImplementedError('Cursor.setinputsizes not supported')
+
+  def setoutputsize(self, size, column):
+    """Raises NotImplementedError."""
+    raise NotImplementedError('Cursor.setoutputsize not supported')
+
+  def _init_delegate(self):
+    self._check_closed()
+    if self._delegate is None:
+      self._delegate = self.connection._delegate.cursor()
+
+  def _check_that_read_query_was_issued(self):
+    self._check_closed()
+    if self._delegate is None:
+      raise ValueError('no read query was issued')
+
+  def _check_closed(self):
+    if self._is_closed:
+      raise ValueError('cursor was closed')
+
+
+class _TransactionConnection(Connection):
+  """PEP 249 Connection object when inside TensorBoard transactions."""
+
+  def __init__(self, connection):
+    super(_TransactionConnection, self).__init__(connection)
+    self.write_queries = []
+
+  def cursor(self):
+    return _TransactionCursor(self)
+
+  def commit(self):
+    raise NotImplementedError('Please return from callback to commit')
+
+  def rollback(self):
+    raise NotImplementedError('Please throw an exception to rollback')
+
+  def close(self):
+    raise NotImplementedError('Connection.close not available in transactions')
+
+  def __enter__(self):
+    raise NotImplementedError('Already in a transaction')
+
+  def __exit__(self, exc_type, exc_value, traceback):
+    pass
+
+
+class _TransactionCursor(Cursor):
+  """PEP 249 Cursor object when inside TensorBoard transactions."""
+
+  FORBIDDEN_PATTERN = re.compile(r'^\s*(ALTER|CREATE|DROP|VACUUM)\b', re.I)
+  WRITE_QUERY_PATTERN = re.compile(r'^\s*(DELETE|INSERT|REPLACE|UPDATE) ', re.I)
```

Comment:
More generally…the only non-write query that we care about is `SELECT`, right? So what do you think about using a whitelist instead of a blacklist—something like (untested) `READ_ONLY_QUERY_PATTERN = re.compile(r'^(?:\s|/\*(?:.(?<!\*/))*?\*/|--[^\n]*?\n)*SELECT\b', re.I | re.S)`?

(Would that we had possessive quantifiers…)